src/mds/Server.cc

```diff
@@ -364,43 +362,55 @@ void Server::handle_client_session(MClientSession *m)
       return;
     }
 
-    session->set_client_metadata(m->client_meta);
-    dout(20) << __func__ << " CEPH_SESSION_REQUEST_OPEN "
-      << session->info.client_metadata.size() << " metadata entries:" << dendl;
-    for (map<string, string>::iterator i = session->info.client_metadata.begin();
-        i != session->info.client_metadata.end(); ++i) {
-      dout(20) << "  " << i->first << ": " << i->second << dendl;
-    }
-
-    // Special case for the 'root' metadata path; validate that the claimed
-    // root is actually within the caps of the session
-    if (session->info.client_metadata.count("root")) {
-      const auto claimed_root = session->info.client_metadata.at("root");
-      // claimed_root has a leading "/" which we strip before passing
-      // into caps check
-      if (claimed_root.empty() || claimed_root[0] != '/' ||
-          !session->auth_caps.path_capable(claimed_root.substr(1))) {
-        derr << __func__ << " forbidden path claimed as mount root: "
-             << claimed_root << " by " << m->get_source() << dendl;
-        // Tell the client we're rejecting their open
-        mds->send_message_client(new MClientSession(CEPH_SESSION_REJECT), session);
-        mds->clog->warn() << "client session with invalid root '" <<
-          claimed_root << "' denied (" << session->info.inst << ")";
-        session->clear();
-        // Drop out; don't record this session in SessionMap or journal it.
-        break;
+    {
+      client_metadata_t client_metadata(m->metadata, m->supported_features);
+      dout(20) << __func__ << " CEPH_SESSION_REQUEST_OPEN metadata entries:" << dendl;
+      dout(20) << "  features: '" << client_metadata.features << dendl;
+      for (auto& p : client_metadata) {
+	dout(20) << "  " << p.first << ": " << p.second << dendl;
       }
-    }
 
-    if (session->is_closed())
-      mds->sessionmap.add_session(session);
+      feature_bitset_t missing_features(CEPHFS_FEATURES_MDS_REQUIRED);
+      missing_features -= client_metadata.features;
+      if (!missing_features.empty()) {
+	mds->send_message_client(new MClientSession(CEPH_SESSION_REJECT), session);
```

Comment:
Can we also give a reason for the rejection so the client can give a useful error message?